CI for torchhead hangs after `TorchTests::test_horovod_join_allgather`

**Environment:**
1. Framework: PyTorch
2. Framework version: Nightly
3. Horovod version: master

**Bug report:**
As observed in https://github.com/horovod/horovod/pull/3558#issuecomment-1148685657, having fixed the build for TF 2.10 nightly, tests for `torchhead` now seem to hang just after `TorchTests::test_horovod_join_allgather`.

Testing locally (Horovod master built with torch nighty), it appears that `test_horovod_join_allreduce` fails for one worker, while the other worker hangs. That's the test right after `..._join_allgather`:
```
test/parallel/test_torch.py::TorchTests::test_horovod_join_allgather PASSED [ 73%]PASSED [ 73%]
test/parallel/test_torch.py::TorchTests::test_horovod_join_allreduce
test/parallel/test_torch.py::TorchTests::test_horovod_join_allreduce FAILED [ 74%]
test/parallel/test_torch.py::TorchTests::test_horovod_join_broadcast
# hang
```
Running `test_horovod_join_allreduce` individually, I get a clear failure message:
```
>               assert torch.allclose(averaged_b, div(tensor_b * (size - 1), size), threshold), \
                    'hvd.join with hvd.allreduce produces incorrect results'
E               AssertionError: hvd.join with hvd.allreduce produces incorrect results
E               assert False
E                +  where False = <built-in method allclose of type object at 0x7fb4076c8840>(tensor([ 32, -34, -37,  37,  14,  21, -29,  27,  41,  31,  21,  48,  24, -29,\n         41,  -1, -34], dtype=torch.int32), tensor([ 32, -34, -38,  37,  14,  21, -30,  27,  41,  31,  21,  48,  24, -29,\n         41,  -1, -34], dtype=torch.int32), 0)
E                +    where <built-in method allclose of type object at 0x7fb4076c8840> = torch.allclose
E                +    and   tensor([ 32, -34, -38,  37,  14,  21, -30,  27,  41,  31,  21,  48,  24, -29,\n         41,  -1, -34], dtype=torch.int32) = <function TorchTests.test_horovod_join_allreduce.<locals>.div at 0x7fb3e4bc6c10>((tensor([ 64, -68, -75,  75,  29,  43, -59,  54,  82,  63,  42,  96,  48, -58,\n         82,  -2, -68], dtype=torch.int32) * (2 - 1)), 2)
```
So we are comparing two tensors and finding them to be unequal:
> tensor([ 32, -34, **-37**,  37,  14,  21, **-29**,  27,  41,  31,  21,  48,  24, -29, 41,  -1, -34], dtype=torch.int32),
> tensor([ 32, -34, **-38**,  37,  14,  21, **-30**,  27,  41,  31,  21,  48,  24, -29, 41,  -1, -34], dtype=torch.int32),

The differences (bold) are +/- 1 and the differing values are both negative. 

One of the tensors has been produced by PyTorch `div(..., 2)` in the test code: https://github.com/horovod/horovod/blob/ba98ad533f8ea30a43d2bf4d3c8064e50109f4ac/test/parallel/test_torch.py#L2827

I think the other tensor has been halved in Horovod code by casting each element to float, multiplying by 0.5, and casting back to integer. This happens in `ScaleBuffer`, called here for example: https://github.com/horovod/horovod/blob/ba98ad533f8ea30a43d2bf4d3c8064e50109f4ac/horovod/common/ops/gloo_operations.cc#L178

Presumably PyTorch may have changed how they round their integer divisions towards zero by default? 

In that case the easiest way to proceed may be setting an explicit `rounding_mode` in the PyTorch unit tests.

--- a/test/parallel/test_torch.py
+++ b/test/parallel/test_torch.py
@@ -42,6 +42,7 @@ sys.path.append(os.path.join(os.path.dirname(__file__), os.pardir, 'utils'))
 
 from common import mpi_env_rank_and_size, skip_or_fail_gpu_test, temppath
 
+_1_12_api = LooseVersion(torch.__version__) >= LooseVersion('1.12.0')
 _1_5_api = LooseVersion(torch.__version__) >= LooseVersion('1.5.0')
 _is_mac = platform.system() == 'Darwin'
 
@@ -2781,6 +2782,8 @@ class TorchTests(unittest.TestCase):
             torch.manual_seed(1234)
 
             def div(t, s):
+                if _1_12_api and dtype in integral_types:
+                    return t.div(s, rounding_mode='trunc')
                 if _1_5_api and dtype in integral_types:
                     return t.floor_divide(s)
                 return t / s

